source/grid/grid_generator_pipe_junction.cc

```diff
@@ -0,0 +1,460 @@
+    // the magnitude one in each of the skeleton directions and span a plane
+    // with the three points we reached.
+    //
+    // The normal vector of this particular plane then describes the edge at
+    // which all pipe segments meet. If we would interpret the bifurcation as a
+    // ball joint, the normal vector would correspond to the polar axis of the
+    // ball.
+    const auto normal = [&]() {
+      constexpr unsigned int               n_pipes = 3;
+      std::array<Point<spacedim>, n_pipes> points;
+      for (unsigned int p = 0; p < n_pipes; ++p)
+        points[p] = bifurcation.first - skeleton_unit[p];
+
+      const auto normal =
+        cross_product_3d(points[1] - points[0], points[2] - points[0]);
+      Assert(normal.norm() > tolerance_length,
+             ExcMessage("Invalid input: all three openings "
+                        "are located on one line."));
+
+      return normal / normal.norm();
+    }();
+
+    // Projections of all skeleton vectors perpendicular to the normal vector,
+    // or in other words, onto the plane described above.
+    const auto skeleton_plane = [&]() {
+      constexpr unsigned int      n_pipes = 3;
+      std::array<vector, n_pipes> skeleton_plane;
+      for (unsigned int p = 0; p < n_pipes; ++p)
+        {
+          skeleton_plane[p] = skeleton[p] - (skeleton[p] * normal) * normal;
+          Assert(std::abs(skeleton_plane[p] * normal) <
+                   tolerance * skeleton_plane[p].norm(),
+                 ExcInternalError());
+          Assert(skeleton_plane[p].norm() > tolerance_length,
+                 ExcMessage("Invalid input."));
+        }
+      return skeleton_plane;
+    }();
+
+    // Create a hyperball domain in 2D that will act as the reference cross
+    // section for each pipe segment.
+    const auto tria_base = []() {
+      Triangulation<dim - 1, spacedim - 1> tria_base;
+      GridGenerator::hyper_ball_balanced(tria_base,
+                                         /*center=*/Point<spacedim - 1>(),
+                                         /*radius=*/1.);
+      return tria_base;
+    }();
+
+    // Now move on to actually build the pipe junction geometry!
+    //
+    // For each pipe segment, we create a separate triangulation object which
+    // will be merged with the parameter triangulation in the end.
+    Assert(tria.n_cells() == 0,
+           ExcMessage("The output triangulation object needs to be empty."));
+    for (unsigned int p = 0; p < n_pipes; ++p)
+      {
+        Triangulation<dim, spacedim> pipe;
+
+        //
+        // Step 1: create unit cylinder
+        //
+        // We create a unit cylinder by extrusion from the base cross section.
+        // The number of layers depends on the ratio of the length of the
+        // skeleton and the minimal radius in the pipe segment.
+        const unsigned int n_slices =
+          1 + std::ceil(skeleton_length[p] /
+                        std::min(openings[p].second, bifurcation.second));
```

Comment:
It felt right to set the number of slices for the extrusion like this. However, users might want to have control over the number of slices. Shall we provide some factor for this as a parameter?